arch/x86: add SMP support (64-bit mode only)

gnuless opened the pull request:
Title says it all.  Add support for up to 4 CPUs (which is the limit currently imposed by the kernel).

Details in the commits.

Commit:
arch/x86: (Intel64) add z_arch_curr_cpu() to enable CONFIG_SMP=y

And set qemu_x86_long board to build with CONFIG_SMP=y by default.
Apparently two benchmark tests - latency_measure and sys_kernel -
do not work with the SMP scheduler, so those tests are disabled.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
kernel/Kconfig: clamp number of CPUs to 4

This maximum is implicit in the kernel support for SMP, e.g.,
kernel/init.c and kernel/smp.c assume CONFIG_MP_NUM_CPUS <= 4.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
drivers/interrupt_controller/loapic.h: add IPI support

Add a simple inline function and some definitions to faciliate
inter-processor interrupts for SMP initialization/synchronization.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: (Intel64) start up on _interrupt_stack, not _exception_stack

Simply for consistency with other platforms.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: elevate prep_c.c to common code

Elevate the previously 32-bit-only z_x86_prep_c() function to common
code, so both 32-bit and 64-bit arches now enter the kernel this way.
Minor changes to prep_c.c to make it build with the SMP scheduler on.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: (Intel64) add missing linker symbols

The linker script was missing symbols that defined the boundaries
of kernel memory segments (_image_rom_end, etc.). These are added
so that core/memmap.c can properly account for those segments.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: factor out common assembly startup code

The 32-bit and 64-bit assembly startup sequences share quite a
bunch of common code, so it's factored out into one file to avoid
repeating ourselves (and potentially falling out of sync).

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: drivers/loapic_intr.c: move local APIC initialization

In the general case, the local APIC can't be treated as a normal device
with a single boot-time initialization - on SMP systems, each CPU must
initialize its own. Hence the initialization proper is separated from
the device-driver initialization, and said initialization is called
from the early startup-assembly code when appropriate.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: (Intel64) clear BSS before entering long mode

This is really just to facilitate CPU bootstrap code between
the BSP and the APs, moving the clear operation out of the way.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: (Intel64) optimize and re-order startup assembly sequence

In some places the code was being overly pedantic; e.g., there is no
need to load our own 32-bit descriptors because the loader's are fine
for our purposes. We can defer loading our own segments until 64-bit.

The sequence is re-ordered to faciliate code sharing between the BSP
and APs when SMP is enabled (all BSP-specific operations occur before
the per-CPU initialization).

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: (Intel64) rename kernel segment constants

There's no need to qualify the 64-bit CS/DS selectors, and the GS and
TR selectors are renamed CPU0_GS and CPU0_TR as they are CPU-specific.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: (Intel64) use per-CPU parameter struct for CPU startup

A new 'struct x86_cpuboot' is created as well as an instance called
'x86_cpuboot[]' which contains per-CPU boot data (initial stack,
entry function/arg, selectors, etc.). The locore now consults this
table to set up per-CPU registers, etc. during early boot.

Also, rename tss.c to cpu.c as its scope is growing.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: alter signature of z_x86_prep_c() function

Take a dummy first argument, so that the BSP entry point (z_x86_prep_c)
has the same signature as the AP entry point (smp_init_top).

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: (Intel64) implement SMP support

Add duplicate per-CPU data structures (x86_cpuboot, tss, stacks, etc.)
for up to 4 total CPUs, add code in locore and z_arch_start_cpu().

The test board, qemu_x86_long, now defaults to 2 CPUs.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

gnuless commented:
There are quite a few sanitycheck failures, though I am quite sure that most (if not all) are related to lingering SMP problems in the test cases and/or a recent regression in the scheduler's support for non-IPI-enabled systems, for which an issue will be filed shortly.

Scheduler IPI support is also coming soon.  It's a very straightforward addition, I'm just being lazy.

gnuless commented:
Force-push for rebase; some commits have been merged in other PRs.

Commit:
arch/x86: limit number of IRQ vectors to 224

Trivial change to the Kconfig: the first 32 vectors are reserved,
so it's not possible to have 256 IRQ vectors. Change max to 224.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: (Intel64) fix conditional assembly in locore.S

was ignoring the rest of the expression, though the effect was
harmless (including unreachable code in some builds).

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: (Intel64) add scheduler IPI support

Add z_arch_sched_ipi() and such to enable scheduler IPIs when SMP.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: (Intel64) do not lock interrupts around irq_offload()

This is the Wrong Thing(tm) with SMP enabled. Previously this
worked because interrupts would be re-enabled in the interrupt
entry sequence, but this is no longer the case.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
drivers/interrupt_controller/ioapic_intr.c: broadcast interrupts

When SMP is enabled, the kernel expects that interrupts be delivered
to all CPUs in the system. Change the I/O APIC RTEs accordingly.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: (Intel64) make z_arch_start_cpu() synchronous

Don't leave z_arch_start_cpu() until the target CPU has been started.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

gnuless commented:
`sanitycheck` is lying a bit here, I see sporadic failures in my both of my build environments, but these failures appear to appear `qemu_x86_64` as much as `qemu_x86_long`, so I do not believe the problem is related to the arch-specific code.

gnuless commented:
@ceolin @andrewboie @andyross @dcpleung @nashif No takers?

andrewboie commented on arch/x86/core/ia32/crt0.S:
```diff
@@ -235,6 +235,10 @@ __csSet:
 	lidt	z_x86_idt		/* load 32-bit operand size IDT */
 
+#ifdef CONFIG_LOAPIC
+	call	z_loapic_enable
```
I'm wondering if we can just put this call in z_x86_prep_c instead of here, since it appears that for both 32 and 64 we just call z_x86_prep_c immediately afterwards anyway.

gnuless replied:
On 64-bit systems, only the BSP lands in the kernel at `z_x86_prep_c()` .. the rest of the processors enter at `smp_init_top()`.  Short story is that we could do what you suggest, and consolidate the code paths in this way, but that means splitting them somewhere else.  It's sorta ugly either way, unfortunately.

andrewboie replied:
Ah OK, let's leave it.

andrewboie commented on arch/x86/core/intel64/locore.S:
```diff
@@ -31,6 +31,17 @@ __start:
 
+	/*
+	 * clear the BSS. note that we do this in 32-bit mode, so
+	 * the BSS must fit entirely in the first 4GB of RAM.
```
nitpick: consider adding a build-time check for this somewhere, although this is an astonishingly unlikely situation.

andrewboie commented on arch/x86/core/intel64/cpu.c:
```diff
@@ -7,6 +7,9 @@
 #include <kernel_structs.h>
+#include <arch/x86/multiboot.h>
+
+extern FUNC_NORETURN void z_x86_prep_c(struct multiboot_info *);
```
probably better to just put this prototype in kernel_arch_func.h.
Not sure if our coding standard allows prototypes without argument names.

gnuless replied:
I've pushed a commit to move the prototype. Fixed.

andrewboie commented on arch/x86/core/intel64/irq.c:
```diff
@@ -92,15 +92,11 @@ int z_arch_irq_connect_dynamic(unsigned int irq, unsigned int priority,
 {
-	u32_t key;
-
-	key = irq_lock();
```
IIRC, the intent here was to serialize setting x86_irq_funcs/args so that concurrent entry to this function didn't clobber stuff. maybe a spinlock here instead?

this API is only used for tests, seems very unlikely in practice though.

gnuless replied:
Well, the proper way to synchronize here has never been settled.  I don't think a spinlock is a good solution, because holding a spinlock should imply that interrupts are disabled on the local core (@andyross can confirm if this is the design intent).  We shouldn't *simulate* an interrupt under those circumstances, either.

On UP systems, the `irq_lock()`/`irq_unlock()` pairing is misleading -- interrupts are NOT disabled in between them, they're only disabled long enough to set the offload function and its argument, and then are re-enabled in the IRQ handler path, only to be briefly disabled again in the IRQ exit path, basically right before `irq_unlock()` is called to re-enable them.

tyrantlizard77 reviewed (approved):
I can only find trivial stuff to complain about, otherwise LGTM.

andrewboie commented:
> -:952: WARNING:NEW_TYPEDEFS: do not add new typedefs
> #952: FILE: arch/x86/include/intel64/kernel_arch_data.h:94:
> +typedef struct x86_cpuboot x86_cpuboot_t;

Gah, our offset generation *requires* typedefs, on a rainy day we ought to revise this. Ignore this.

The rest should be addressed, trivial formatting stuff.

Commit:
arch/x86: declare z_x86_prep_c() in kernel_arch_func.h

And remove the ad hoc prototype in cpu.c for Intel64.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

gnuless commented:
> The rest should be addressed, trivial formatting stuff.

Done.

gnuless commented:
> Gah, our offset generation _requires_ typedefs, on a rainy day we ought to revise this.

I think we go about this the hard way.  A more conventional method this is to compile a program that includes all the requisite headers and then spits out a header file with manifest constants `#define`d.  I'm really not sure why we go through all this trouble to inject symbols into the object files, perhaps there's a good reason for it that I'm not aware of ...

Commit:
arch/x86: (Intel64) initialize TSS interrupt stack from cpuboot[]

In non-SMP MP situations, the interrupt stacks might not exist, so
do not assume they do. Instead, initialize the TSS IST1 from the
cpuboot[] vector (meaning, on APs, the stack from z_arch_start_cpu).
Eliminates redundancy at the same time.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
arch/x86: do not assume MP means SMP

It's possible to have multiple processors configured without using the
SMP scheduler, so don't make definitions dependent on CONFIG_SMP.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

andrewboie commented:
>  A more conventional method this is to compile a program that includes all the requisite headers and then spits out a header file with manifest constants `#define`d. I'm really not sure why we go through all this trouble to inject symbols into the object files, perhaps there's a good reason for it that I'm not aware of ...

All it's used for AFAIK, is to compute sizeof() various structs, and offsets of struct members, and emit a header file with these values, for use within assembly code. It needs typedefs due to the way it generates names but that could be adjusted without too much hassle I think.

gnuless commented:
> All it's used for AFAIK, is to compute sizeof() various structs, and offsets of struct members, and emit a header file with these values, for use within assembly code.

That's what you'd expect, but instead we use a mess of macros (imagine that) to generate absolute symbols in the object files.  If we generated headers, it'd actually be better, because then the assembler could take advantage of the fact that they're constants.  Consider sequences like

```
movl $_kernel, %esi
movl _kernel_offset_to_current(%esi), %esi
```

This should simply be 

```
movl _kernel+_kernel_offset_to_current, %esi
```

because "address of an object plus an offset" can be represented in an object file, but "the sum of two symbols" can't be.

Anyway, I digress.  As you said, rainy-day stuff.

Commit:
tests/kernel/smp: make 64-bit clean

Just housekeeping around the casting between void * arguments to
thread functions and integer types.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

andrewboie commented:
> That's what you'd expect, but instead we use a mess of macros (imagine that) to generate absolute symbols in the object files. If we generated headers, it'd actually be better, because then the assembler could take advantage of the fact that they're constants. 

No kidding! (I never dived into the details on how it worked). I agree, seems needlessly complicated given the problem it's supposed to solve. Worth an enhancement GH ticket I think to clean it up someday.

Aside from the two build errors in CI I think this PR is ready to go in.

nashif commented:
> No kidding! (I never dived into the details on how it worked). I agree, seems needlessly complicated given the problem it's supposed to solve. Worth an enhancement GH ticket I think to clean it up someday.

agree this needs to be enhanced, the offset generation step is legacy and is overly complicates the build process for now good reason. Lets looks into solving this finally.
